README.md
# DEPRECATION WARNING
This userscript is deprecated. Please install <A HREF="https://github.com/TomONeill/seriesfeed-transporter">Seriesfeed Transporter</A> instead.

# Seriesfeed Importer
Import your favourites from other sites like Bierdopje.com to Seriesfeed.com
<BR/>
Enjoy.
<BR/><BR/>
Version <strong>2.4.0</strong>

<A HREF="https://github.com/TomONeill/Seriesfeed-Importer/raw/master/SeriesfeedImport.user.js">INSTALL</A>

# Information
With this userscript, you can import your favourites from Bierdopje.com to Seriesfeed.com.<BR />
Your username will be filled in automatically if you are logged onto Bierdopje.com. You can use any username you like, however. After clicking the Import button, your favourites are being imported into Seriesfeed. In the Details table, rows are being given a colour if there is something about that series. Blue rows will indicate you most likely already have that series as a favourite on Seriesfeed. Yellow ones indicate the series does not exist on Seriesfeed yet and red ones indicate that an error has happened while connecting Bierdopje or Seriesfeed. You can always try again to see if it works the second time.

# Screenshots
<img src="https://raw.githubusercontent.com/TomONeill/Seriesfeed-Importer/master/Screenshots/v1.1-1.png" alt="Version 2.2" width="250px" />
<img src="https://raw.githubusercontent.com/TomONeill/Seriesfeed-Importer/master/Screenshots/v2.0-2.png" alt="Version 2.2" width="250px" />

# Changelog
<A HREF="https://raw.githubusercontent.com/TomONeill/Seriesfeed-Importer/master/Changelog.txt">View changelog</A>

# Other userscripts
Check out other userscripts for Seriesfeed by me:<BR/>
<ul>
    <li><A HREF="https://github.com/TomONeill/seriesfeed-transporter">Seriesfeed Transporter</A>: Import and export your favourites and time wasted.</li>
    <li><A HREF="https://github.com/TomONeill/Seriesfeed-Move">Seriesfeed Move</A>: drag and drop frontpage items to your liking.</li>
    <li><A HREF="https://github.com/TomONeill/seriesfeed-episode-inverter">Seriesfeed Episode Inverter</A>: Choose whether episodes are sorted ascending or descending.</li>
    <li><A HREF="https://github.com/TomONeill/Seriesfeed-Season-Separator">Seriesfeed Season Separator</A>: Draws a thick line between seasons.</li>
</ul>

# Development
If you want to help with the development, you can. From now on I'll be working in the `src/Seriesfeed-Importer-TS` directory, in which a TypeScript project is located. You can use this for development. To setup the dev in combination with Tampermonkey/other, do the following:
1. Open the `src/Seriesfeed-Importer-TS` folder in your favourite IDE.
2. Open CMD, PowerShell or the terminal (VS Code tip: use CTRL + \`).<br/>
2.5: If not already, `cd` to this repository.
3. Type: `tsc --watch` to run the TypeScript compiler. The `tsconfig.json` file will determine the location of the output.
4. Add a userscript with the following:
```
	// ==UserScript==
	// @name         Seriesfeed Importer
	// @namespace    https://www.seriesfeed.com
	// @version      3.0-dev
	// @description  Import your favourites and Time Wasted from Bierdopje.com.
	// @match        https://*.seriesfeed.com/*
	// @grant        unsafeWindow
	// @grant        GM_xmlhttpRequest
	// @connect      www.bierdopje.com
	// @connect      www.imdb.com
	// @domain       www.bierdopje.com
	// @domain       www.imdb.com
	// @require      https://code.jquery.com/jquery-3.2.1.min.js
	// @require      file://{YOUR_CLONE}\Seriesfeed-Importer\src\Seriesfeed-Importer-TS\SeriesfeedImport.user.js
	// @author       Tom
	// @copyright    2016 - 2017, Tom
	// ==/UserScript==
	/* jshint -W097 */
	/* global $, GM_xmlhttpRequest, Promise, console */
	'use strict';
```
5. Make sure you have ticked the box `Allow access to file URLs` for the Tampermonkey extension.

Any change will compile (some would say transpile) and a browser refresh will do the rest.

# Donate
If you like my work so much you feel like doing something nice for me, a complete stranger of the internet, you can.<BR />
<A HREF="https://www.paypal.me/TomONeill">Donate here</A>.


Changelog.txt
Seriesfeed Importer Changelog

Version 2.4.0 (15 November 2017)
Updated jQuery library + implementation.
Experimented with GM 4 =(

Version 2.3.1 (17 October 2017)
Fixed status messages per imported series.

Version 2.3 (2 October 2017)
Added compatibility for Seriesfeed v2.5.

Version 2.2 (20 April 2016)
Added colours to identify the status of a request.
Improved error handling which should fix the freezing bug.
Eased down the queueing mechanism some more due to Firefox issues.

Version 2.1.1 (18 April 2016)
Eased down the queueing mechanism.

Version 2.1 (17 April 2016)
Created a queueing type of requesting mechanism.

Version 2.0 (14 April 2016)
Synchronously adding favourites instead of mass-requesting favourites (fixes sudden freeze).
Replaced HTML5 progress-bar with Bootstrap's version.
Fixed table-jumping by introducing colgroup.
Auto-get Bierdopje username.
Rewritten a lot of code.

Version 1.3 (10 April 2016)
Namechange: Seriesfeed Import -> Seriesfeed Importer.
Fixed: Unknown series results in empty series name.
Fixed: Menuitem doesn't show up on subdomain pages of Seriesfeed.
Details on importing is now open by default.
Changed header on the detais table.
Links on the detais table are now opened on another tab.

Version 1.2 (10 April 2016)
Replaced the name/slug search by TVDb matching.
Fixed some jshint errors.
Fixed indentation.
Adds gitignore.

Version 1.1 (3 April 2016)
Updated UI of the platform selection screen.

Version 1.0 (3 April 2016)
Removed the need of an API.
Added UI.

Version 0.1 (2 April 2016)
First working draft.


old-versions/SeriesfeedImport-v2.2.user.js
// ==UserScript==
// @name         Seriesfeed Importer
// @namespace    http://www.seriesfeed.com
// @version      2.2
// @description  Allows you to import your favourites from Bierdopje.com.
// @updateURL 	 https://github.com/TomONeill/Seriesfeed-Importer/raw/master/SeriesfeedImport.user.js
// @match        http://*.seriesfeed.com/*
// @run-at       document-start
// @grant        unsafeWindow
// @grant        GM_xmlhttpRequest
// @connect      www.bierdopje.com
// @domain       www.bierdopje.com
// @require      http://code.jquery.com/jquery-1.12.2.min.js
// @author       Tom
// @copyright    2016+, Tom
// ==/UserScript==
/* jshint -W097 */
/* global $, GM_xmlhttpRequest, Promise, console */
'use strict';

$(function() {
	// Add menu item to navigator.
	$('.nav .dropdown .dropdown-menu:eq(2)').append('<li><a href="/series/import/">Importeren</a></li>');

	if (window.location.href === "http://www.seriesfeed.com/series/import/") {
		var wrapper    = $('.wrapper').addClass('bg').removeClass('show padding');
		var container  = $('.wrapper .container').html('').addClass("content");
		var selector   = $(document.createElement("div")).addClass("platformSelector");
		var card       = $(document.createElement("div")).addClass("wow fadeInUp cardStyle cardForm formBlock animated");
		var importHead = $(document.createElement("h2")).append('Series importeren');
		var cardInner  = $(document.createElement("div")).addClass("cardFormInner");
		var platform   = $(document.createElement("p")).append('Kies een platform:');

		container.append(selector);
		selector.append(card);
		card.append(importHead);
		card.append(cardInner);
		cardInner.append(platform);

		var bierdopje = platformFactory("Bierdopje.com", "http://cdn.bierdopje.eu/g/layout/bierdopje.png", "bierdopje/", "#3399FE");
		bierdopje.addClass('wow fadeInLeft cardStyle cardForm formBlock animated');
		platform.after(bierdopje);
	}

	if (window.location.href === "http://www.seriesfeed.com/series/import/bierdopje/") {
		var head = $('.col-md-12 h1').html('');
		var p    = $('.col-md-12 p').html('');
		
		var css = '<style>'
		        + '    .progress {'
		        + '        margin-top: 10px;'
		        + '        margin-bottom: 0px;'
		        + '    }'
		        + '    '
		        + '    .progress-bar {'
                + '        background: #447C6F;'
		        + '    }'
                + '    '
		        + '    tr.row-error {'
		        + '        background-color: rgba(255, 0, 0, 0.15);'
		        + '    }'
                + '    '
		        + '    tr.row-warning {'
		        + '        background-color: rgba(255, 231, 150, 0.43);'
		        + '    }'
                + '    '
		        + '    tr.row-info {'
		        + '        background-color: rgba(240, 248, 255, 1.00);'
		        + '    }'
		        + '</style>';
		$('body').append(css);

		var formElement   = $(document.createElement("div"));
		var usernameInput = $('<div><input type="text" id="username" class="form-control" placeholder="Gebruikersnaam" /></div>');
		var submitInput   = $('<div><input type="button" id="fav-import" class="btn btn-success btn-block" value="Favorieten Importeren" /></div>');
		var bottomPane    = $('<div class="blog-left"></div>');
		var detailsTable  = $('<table class="table table-hover responsiveTable favourites stacktable large-only" id="details">');
		var colGroup      = $('<colgroup><col width="15%"><col width="35%"><col width="50%"></colgroup>');
		var detailsHeader = $('<tr><th style="padding-left: 30px;">Id</th><th>Serie</th><th>Status</th></tr>');
		var showDetails   = $('<div class="blog-content" id="details-content"><input type="button" id="show-details" class="btn btn-block" value="Details" /></div>');
		
		formElement.addClass('blog-left wow fadeInUp cardStyle cardForm formBlock animated');
		bottomPane.addClass('wow fadeInLeft cardStyle animated');
		detailsTable.addClass('wow fadeInLeft cardStyle animated');
		formElement.css('padding', '10px');

		head.append('Series importeren - Bierdopje.com');
		p.append('Voer je gebruikersnaam in en klik op de knop "Favorieten Importeren"');

		formElement.append(usernameInput);
		formElement.append(submitInput);
		p.after(formElement);
		detailsTable.append(colGroup);
		detailsTable.append(detailsHeader);
		bottomPane.append(showDetails);
		showDetails.append(detailsTable);
		
		getCurrentBierdopjeUsername().then(function(username) {
			$('#username').attr('value', username);
		});

		$("#fav-import").click(function() {
			var favImportBtn = $(this);
			var outerProgress = $('<div class="progress"></div>');
			var progressBar   = $('<div class="progress-bar progress-bar-striped active"></div>');
			
			favImportBtn.prop('disabled', true).attr('value', "Bezig met importeren...");
			outerProgress.append(progressBar);
			formElement.append(outerProgress);
			formElement.after(bottomPane);

			var username = $('#username').val();
			var favourites = $('#details');

			$("#show-details").click(function() {
				detailsTable.toggle();
			});
			
			getAllBierdopjeFavouritesByUsername(username).then(function(bdFavouritesPageData) {
				var bdFavouritesData = $('<div></div>').html(bdFavouritesPageData.responseText);
				var bdFavouriteLinks = bdFavouritesData.find('.content').find('ul').find('li').find('a');
				var bdFavouritesLength = bdFavouriteLinks.length;

				function getBierdopjeFavourite(index) {
					return new Promise(function(resolve) {
						var bdShowName = $(bdFavouriteLinks[index]).text();
						var bdShowSlug = $(bdFavouriteLinks[index]).attr('href');
						var bdShowUrl = 'http://www.bierdopje.com/shows';

						getTVDBIdByBierdopjeShowSlug(bdShowSlug).then(function(tvdbId) {
							getSeriesfeedShowDataByTVDBId(tvdbId).success(function(sfShowData) {
								var sfSeriesId   = sfShowData.id;
								var sfSeriesName = sfShowData.name;
								var sfSeriesSlug = sfShowData.slug;
								var sfSeriesUrl  = 'http://www.seriesfeed.com/series/';

								addSeriesfeedFavouriteByShowId(sfSeriesId).success(function(result) {
									var resultStatus = result.status;
									var item = "<tr></tr>";
									var status = "-";
									var showUrl = sfSeriesUrl + sfSeriesSlug;
									
									if (sfSeriesId === -1) {
										sfSeriesId = "Onbekend";
									}
									
									if (!sfSeriesName) {
										showUrl = bdShowUrl + bdShowSlug;
										sfSeriesName = bdShowName;
									}

									if (resultStatus === "success") {
										status = "Toegevoegd als favoriet.";
										item = '<tr><td>' + sfSeriesId + '</td><td><a href="' + showUrl + '" target="_blank">' + sfSeriesName + '</a></td><td>' + status + '</td></tr>';
									} else if (resultStatus === "failed" && sfSeriesId === "Onbekend") {
										status = '<a href="' + sfSeriesUrl + 'voorstellen/" target="_blank">Deze serie staat nog niet op Seriesfeed.</a>';
										item = '<tr class="row-warning"><td>' + sfSeriesId + '</td><td><a href="' + showUrl + '" target="_blank">' + sfSeriesName + '</a></td><td>' + status + '</td></tr>';
									} else {
										status = "Deze serie is mogelijk al een favoriet en is dus ongewijzigd.";
										item = '<tr class="row-info"><td>' + sfSeriesId + '</td><td><a href="' + showUrl + '" target="_blank">' + sfSeriesName + '</a></td><td>' + status + '</td></tr>';
									}
									
									favourites.append(item);

									var progress = (index/bdFavouritesLength) * 100;
									progressBar.css('width', Math.round(progress) + "%");

									resolve();
								}).error(function(error) {
									console.log("Could not connect to Seriesfeed to favourite series " + sfSeriesName + " with id " + sfSeriesId + ".");
									var status = "Kon deze serie niet als favoriet instellen.";
									var item = '<tr class="row-error"><td>' + sfSeriesId + '</td><td><a href="' + sfSeriesUrl + sfSeriesSlug + '" target="_blank">' + sfSeriesName + '</a></td><td>' + status + '</td></tr>';
									favourites.append(item);

									var progress = (index/bdFavouritesLength) * 100;
									progressBar.css('width', Math.round(progress) + "%");
									
									resolve();
								});
							}).error(function(error) {
								console.log("Could not connect to Seriesfeed to convert TVDB id " + tvdbId + ".");
								var status = 'Kon geen verbinding met Seriesfeed maken om het id op te halen.</a>';
								var item = '<tr class="row-error"><td>Onbekend</td><td><a href="' + bdShowUrl + bdShowSlug + '" target="_blank">' + bdShowName + '</a></td><td>' + status + '</td></tr>';
								favourites.append(item);

								var progress = (index/bdFavouritesLength) * 100;
								progressBar.css('width', Math.round(progress) + "%");

								resolve();
							});
						}, function(error) {
							console.log("Could not connect to Bierdopje to get the TVDB of " + bdShowName + ".");
							var status = 'Deze serie kan niet gevonden worden op Bierdopje.';
							var item = '<tr class="row-error"><td>Onbekend</td><td><a href="' + bdShowUrl + bdShowSlug + '" target="_blank">' + bdShowName + '</a></td><td>' + status + '</td></tr>';
							favourites.append(item);

							var progress = (index/bdFavouritesLength) * 100;
							progressBar.css('width', Math.round(progress) + "%");

							resolve();
						});
					});
				}
				
				var MAX_ASYNC_CALLS = 3;
				var current_async_calls = 0;

				Promise.resolve(1).then(function loop(i) {
					if (current_async_calls < MAX_ASYNC_CALLS) {
						if (i < bdFavouritesLength) {
							current_async_calls += 1;
							getBierdopjeFavourite(i).then(function() {
								current_async_calls -= 1;
							});
							return loop(i + 1);
						}
					} else {
						return new Promise(function(resolve) {
							setTimeout(function() {
								resolve(loop(i));
							}, 80);
						});
					}
				}).then(function() {
					function checkActiveCalls() {
						if(current_async_calls === 0) {
							favImportBtn.prop('disabled', false);
							favImportBtn.attr('value', "Favorieten Importeren");
							outerProgress.removeClass('progress');
							progressBar.replaceWith("Importeren voltooid.");
						} else {
							setTimeout(checkActiveCalls, 80);
						}
					}

					checkActiveCalls();
				}).catch(function(error) {
					console.log("Unknown error: ", error);
				});
			}, function(error) {
                console.log("Could not connect to Bierdopje to get your favourites.");
				window.alert('Kan geen favorieten vinden voor ' + username + '. Dit kan komen doordat de gebruiker niet bestaat, geen favorieten heeft of er is iets mis met je verbinding.');
			});
		});
	}

	function getCurrentBierdopjeUsername() {
		return new Promise(function(resolve, reject) {
			GM_xmlhttpRequest({
				method: "GET",
				url: "http://www.bierdopje.com/stats",
				onload: function(bdStatsPageData) {
					var bdStatsData  = $('<div></div>').html(bdStatsPageData.responseText);
					var username = bdStatsData.find('#userbox_loggedin').find('h4').html();
					resolve(username);
				},
				onerror: function(error) {
				    reject(error);
				}
			});
		});
	}

	function platformFactory(name, image, url, colour) {
		// Element declaration
		var portfolio = $(document.createElement("div"));
		var a         = $(document.createElement("a"));
		var wrapper   = $(document.createElement("div"));
		var hover     = $(document.createElement("div"));
		var img       = $(document.createElement("img"));
		var info      = $(document.createElement("div"));
		var title     = $(document.createElement("div"));
		var h4        = $(document.createElement("h4"));

		// Adding classes
		portfolio.addClass("portfolio mix_all");
		wrapper.addClass("portfolio-wrapper cardStyle");
		hover.addClass("portfolio-hover");
		info.addClass("portfolio-info");
		title.addClass("portfolio-title");

		// Styling
		portfolio.css({
			'display': 'inline-block',
			'width': '100%'
		});
		hover.css({
			'text-align': 'center',
			'background': colour
		});

		// Data binding
		a.attr('href', url);
		img.attr('src', image).attr('alt', name);
		h4.append(name);

		// Element binding
		portfolio.append(a);
		a.append(wrapper);
		wrapper.append(hover);
		hover.append(img);
		wrapper.append(info);
		info.append(title);
		title.append(h4);

		return portfolio;
	}
	
	function getAllBierdopjeFavouritesByUsername(username) {
		return new Promise(function(resolve, reject) {
			GM_xmlhttpRequest({
				method: "GET",
				url: "http://www.bierdopje.com/users/" + username + "/shows",
				onload: function(response) {
                    resolve(response);
				},
				onerror: function(error) {
				    reject(error);
				}
			});
		});
	}
	
	function getTVDBIdByBierdopjeShowSlug(showSlug) {
		return new Promise(function(resolve, reject) {
			GM_xmlhttpRequest({
				method: "GET",
				url: "http://www.bierdopje.com" + showSlug,
				onload: function(bdFavouritePageData) {
					var bdFavouriteData = $('<div></div>').html(bdFavouritePageData.responseText);
					var tvdbId = bdFavouriteData.find('a[href^="http://www.thetvdb.com"]').html();
                    resolve(tvdbId);
				},
				onerror: function(error) {
				    reject(error);
				}
			});
		});
	}

	function getSeriesfeedShowDataByTVDBId(tvdbId) {
		return $.ajax({
			type: "POST",
			url: "/ajax.php?action=getShowId",
			data: { tvdb_id: tvdbId },
			dataType: "json"
		});
	}

	function addSeriesfeedFavouriteByShowId(showId) {
		return $.ajax({
			type: "POST",
			url: "/ajax.php?action=updateFavourite",
			data: {
				series: showId,
				type: 'favourite',
				selected: '0'
			},
			dataType: "json"
		});
	}
});